### ISSUE TITLE
Keycloak SAML

### ISSUE THREAD

**Issue by mrajashree** [rancher/rancher#9286]
Changes in Rancher saml
1. For pingFederate we did two changes to the saml library, one was modifying the metadata to add two fields (NameIDPolicy and Destination), and the other was changing the format of NameIDPolicy in the AuthnRequest. The first change (metadata) throws an error for keycloak, and is not needed for pingFederate, I tested ping-rancher connection by removing this change. So the metadata change in the saml library can be removed.
2. Rancher service provider (SP) metadata is generated only after saving all of Identity Provider attributes, key and cert, and the Identity Provider's metadata. Keycloak needs the SP metadata before creating a connection. It doesn't generate the Identity Provider metadata until the connection is created. The Identity Provider's metadata and attributes aren't included in Rancher SP's metadata. So we can make the SP metadata available before saving the IdP's metadata.

**Comment by prachidamle** [comment]
Note, before we generate the SP metadata, we probably need the key and the cert uploaded, if not the rest.

**Comment by mrajashree** [comment]
@prachidamle Like testshib or amazon aws keep their metadata always available, can we do the same with a pre-existing cert?

**Comment by mrajashree** [comment]
Keycloak IdP can be configured even without rancher's metadata, by manually entering fields, so the 2 step config save isn't required. Other than that `Save` and `Test` can't be done in one call, since `Save` is an API call to save config but `Test` is the actual authentication step.
cc @prachidamle

**Comment by deniseschannon** [comment]
Available in v1.6.8-rc1

**Comment by markrey** [comment]
I want to configure it with my local Keycloak server. How can i add it from Access Control page?
